nowcasting_utils/models/losses/FocalLoss.py

```diff
@@ -0,0 +1,100 @@
+import numpy as np
+import torch
+from torch import nn as nn
+
+
+class FocalLoss(nn.Module):
+    def __init__(
+        self,
+        apply_nonlin=None,
+        alpha=None,
+        gamma=2,
+        balance_index=0,
+        smooth=1e-5,
+        size_average=True,
+    ):
+        """
+        Copied from: https://github.com/Hsuxu/Loss_ToolBox-PyTorch/blob/master/FocalLoss/FocalLoss.py
+
+        This is a implementation of Focal Loss with smooth label cross entropy supported which is proposed in
+        'Focal Loss for Dense Object Detection. (https://arxiv.org/abs/1708.02002)'
+        Focal_Loss= -1*alpha*(1-pt)*log(pt)
+
+        Args:
+            apply_nonlin: nn.Module, a non-linearity to apply to the logits
+            alpha: (tensor) 3D or 4D the scalar factor for this criterion
+            gamma: (float,double) gamma > 0 reduces the relative loss for well-classified examples (p>0.5) putting more
+                    focus on hard misclassified example
+            balance_index: balance class index, should be specific when alpha is float
+            smooth: (float,double) smooth value when cross entropy
+            size_average: (bool, optional) By default, the losses are averaged over each loss element in the batch.
+        """
+        super(FocalLoss, self).__init__()
+        self.apply_nonlin = apply_nonlin
+        self.alpha = alpha
+        self.gamma = gamma
+        self.balance_index = balance_index
+        self.smooth = smooth
+        self.size_average = size_average
+
+        if self.smooth is not None:
+            if self.smooth < 0 or self.smooth > 1.0:
+                raise ValueError("smooth value should be in [0,1]")
+
+    def forward(self, logit, target):
+        if self.apply_nonlin is not None:
+            logit = self.apply_nonlin(logit)
+        num_class = logit.shape[1]
+
+        if logit.dim() > 2:
+            # N,C,d1,d2 -> N,C,m (m=d1*d2*...)
+            logit = logit.view(logit.size(0), logit.size(1), -1)
+            logit = logit.permute(0, 2, 1).contiguous()
+            logit = logit.view(-1, logit.size(-1))
+        target = torch.squeeze(target, 1)
+        target = target.view(-1, 1)
+        #
+        alpha = self.alpha
+
+        if alpha is None:
+            alpha = torch.ones(num_class, 1)
+        elif isinstance(alpha, (list, np.ndarray)):
+            assert len(alpha) == num_class
+            alpha = torch.FloatTensor(alpha).view(num_class, 1)
+            alpha = alpha / alpha.sum()
+        elif isinstance(alpha, float):
+            alpha = torch.ones(num_class, 1)
+            alpha = alpha * (1 - self.alpha)
+            alpha[self.balance_index] = self.alpha
+
+        else:
+            raise TypeError("Not support alpha type")
+
+        if alpha.device != logit.device:
+            alpha = alpha.to(logit.device)
```

Comment:
Whilst the Lightning docs discourage `to(device)`, this code is probably OK because it feels functionally equivalents to PL's recommendation of `alpha = alpha.type_as(logit)`